In [1]:
import numpy as np

## Q5

In [2]:
def hinge_loss(y, y_hat):
    return max((1 - y * y_hat), 0)

def gradient_w(X_i, y_i, w, hinge):
    return w + C * (-2) * X_i * y_i * hinge

def gradient_b(X_i, y_i, hinge):
    return C * (-2) * y_i * hinge

In [3]:
def SGD_SVM(X, y, w, b, max_pass, step_size):
    for t in range(max_pass):
        n = len(X)
        for i in range(n):
            y_hat = np.dot(np.transpose(w), X[i]) + b
            hinge = hinge_loss(y[i][0], y_hat)
            if y[i][0] * y_hat <= 1:
                w = w - step_size * gradient_w(X[i], y[i][0], w, hinge)
                b = b - step_size * gradient_b(X[i], y[i][0], hinge)
            w = 1 / (1 + step_size) * w
    return w, b

## Q6

In [4]:
# some inputs
X = np.array([[1,2],[2,1],[3,1],[3,2]])
y = np.array([[1],[1],[-1],[-1]])
w0 = np.array([0,0])
b0 = 0
max_pass= 50000
step_size = 0.0000001
C = 10000

w, b = SGD_SVM(X, y, w0, b0, max_pass, step_size)

In [5]:
C

10000

In [6]:
w

array([-1.99805291e+00, -7.34489503e-04])

In [7]:
b

4.995740369998062

From the result, we know that w = [-1.99805291e+00, -7.34489503e-04] is very close to [-2, 0] and b = 4.995740369998062 is very close to 5.

By using C = 10000, I can recover the hard-margin SVM solution in Q1.